src/runtime/HalideRuntime.h

```diff
@@ -523,67 +525,157 @@ struct halide_type_t {
+
+#ifdef __cplusplus
+    halide_dimension_t() : min(0), extent(0), stride(0), flags(0) {}
+    halide_dimension_t(int32_t m, int32_t e, int32_t s, uint32_t f = 0) :
+        min(m), extent(e), stride(s), flags(f) {}
+
+    bool operator==(const halide_dimension_t &other) const {
+        return (min == other.min) &&
+            (extent == other.extent) &&
+            (stride == other.stride) &&
+            (flags == other.flags);
+    }
+
+    bool operator!=(const halide_dimension_t &other) const {
+        return !(*this == other);
+    }
+#endif
+} halide_dimension_t;
 
 /**
  * The raw representation of an image passed around by generated
  * Halide code. It includes some stuff to track whether the image is
  * not actually in main memory, but instead on a device (like a
  * GPU). */
-typedef struct buffer_t {
+typedef struct halide_buffer_t {
     /** A device-handle for e.g. GPU memory used to back this buffer. */
-    uint64_t dev;
+    uint64_t device;
+
+    /** The interface used to interpret the above handle. */
+    const halide_device_interface_t *device_interface;
 
     /** A pointer to the start of the data in main memory. In terms of
      * the Halide coordinate system, this is the address of the min
      * coordinates (defined below). */
     uint8_t* host;
 
-    /** The size of the buffer in each dimension. */
-    int32_t extent[4];
-
-    /** Gives the spacing in memory between adjacent elements in the
-    * given dimension.  The correct memory address for a load from
-    * this buffer at position x, y, z, w is:
-    * host + elem_size * ((x - min[0]) * stride[0] +
-    *                     (y - min[1]) * stride[1] +
-    *                     (z - min[2]) * stride[2] +
-    *                     (w - min[3]) * stride[3])
-    * By manipulating the strides and extents you can lazily crop,
-    * transpose, and even flip buffers without modifying the data.
-    */
-    int32_t stride[4];
-
-    /** Buffers often represent evaluation of a Func over some
-    * domain. The min field encodes the top left corner of the
-    * domain. */
-    int32_t min[4];
-
-    /** How many bytes does each buffer element take. This may be
-    * replaced with a more general type code in the future. */
-    int32_t elem_size;
-
-    /** This should be true if there is an existing device allocation
-    * mirroring this buffer, and the data has been modified on the
-    * host side. */
-    HALIDE_ATTRIBUTE_ALIGN(1) bool host_dirty;
-
-    /** This should be true if there is an existing device allocation
-    mirroring this buffer, and the data has been modified on the
-    device side. */
-    HALIDE_ATTRIBUTE_ALIGN(1) bool dev_dirty;
-
-    // Some compilers will add extra padding at the end to ensure
-    // the size is a multiple of 8; we'll do that explicitly so that
-    // there is no ambiguity.
-    HALIDE_ATTRIBUTE_ALIGN(1) uint8_t _padding[10 - sizeof(void *)];
-} buffer_t;
+    /** flags with various meanings. */
```

Comment:
Ah, I see, these values are the index of the bit, not the flag value.

I think it might be better to set the flags to the value of the flag itself. I would probably have done `buf.flags |= halide_buffer_t::flag_host_dirty, assuming that the flag is the value of the flag, not the index of the bit.

If nothing else, this uses a different convention to other flag sets we have in Halide (e.g. Target flags).